README.md
# Slice-Library
[![license](https://img.shields.io/github/license/mashape/apistatus.svg?style=flat-square)]()

Slice-Library is a CodeIgniter library that simulates Laravel's Blade templating system! Slice-Library is also compatible with Modular Extensions - HMVC; saves the compiled template in cache and is easy to use and install.

## Features

+ Requires nearly zero configuration!
+ Easy to install and use.
+ Helps you organize your views folder.
+ 30 directives to use!
+ 89 helper functions ~(eleven required, kkk)~!
+ Does not restrict you from using plain PHP code in your views.
+ Easy to learn and to get used to.
+ Caches files until they are modified!
+ Handles complex language string lines.
+ Provides you clear, thorough documentation.

## Requirements

- PHP version 5.6 or newer is recommended.
- CodeIgniter version 2.2.1 or newer.

## Instalation

To use the Slice-Library you must first copy the file located in `application/config/slice.php` to your own `application/config/` folder. Then, edit this file according to your configurations.
Now, copy the file `application/libraries/Slice.php` to your own `application/libraries/` folder.
Finally, make sure the folder `application/cache/` has a **`0664`** permission to save the compiled templates the library will produce.

That's all! Have fun!

## Loading Slice-Library

Your can load the Slice-Library as you load any other library in CodeIgniter:

```php
$this->load->library('slice');
```

### Slice Helper functions

Since version 1.3.0 Slice-Library comes with a helper file with 89 functions! This functions will help you a lot. Some of the are:

+ `app()` - Returns any library instance
+ `decrypt()` - Decrypts a given string
+ `e()` - Escapes HTML entities in a string
+ `email()` - Sends an email
+ `helper()` - Loads any CodeIgniter helper
+ `input()` - Retrieves input item from the request
+ `query()` - Executes a query string
+ `session()` - Gets or sets session values
+ `str_contains()` - Determines if a given string contains a given substring
+ `validator()` - Validates post fields with CodeIgniter Form Validation Class
+ `view()` - Get the evaluated view contents for the given view

In the future a full list of helpers and it's definitions and how to use them will be available.

To use them, make sure the variable `enable_helper` is set `TRUE` in the `config/slice.php` file to load the helpers automatically or load the file as you would do with any CodeIgniter helper.

### Other Libraries and Helpers

Within Slice you can optionally load any Library and/or Helper! This is very useful if you use a library and/or a helper to handle assets, for example, and you don't want to put them in the `config/autoload.php` file.

So, to autoload any library and/or helper set the variable `enable_autoload` in the `config/slice.php` file to `TRUE`. Then add the library you want to load in the config array `libraries` and the helpers you want to load in the config array `helpers`.

This way, your libraries and helpers will be loaded automatically!

### Creating Views

Slice-Library has its own `view` method to display your HTML pages. So, to show a view you can do the following:

```php
$this->slice->view('page', ['name' => 'GustMartins']);
```

It is important to remember that your view files **MUST** have `.slice.php` as extension. But you can change that in your `application/config/slice.php` file.

As you can see, the first argument passed to the `view` method corresponds to the name of the view file in your `views` directory. The second argument is an array of data that should be made available to the view. In this case, we are passing the `name` variable, which is displayed in the view using the syntax explained bellow.

Of course, views may also be nested within sub-directories of you `views` directory. "Dot" notation may be used to reference nested views. For example:

```php
$this->slice->view('user.profile', $data);
```

### Determining If A View Exists

If you need to determine if a view exists, you may use the `exists` method. The `exists` method will return *TRUE* if the view exists:

```php
$this->slice->exists('user.email');
```

### Passing Data To Views

As you saw in the previous examples, you may pass an array of data to views:

```php

$this->slice->view('view', ['name' => 'GustMartins']);

```

When passing information in this manner, `$data` should be an array of key/value pairs. Inside your view, you can then access each value using its corresponding key, such as `<?php echo $key; ?>`. As an alternative to passing a complete array of data to the `view` method, you may use the `with` method to add individual pieces of data, or the `set` method to add an array of key/value pairs of data.

```php
$this->slice->with('name', 'GustMartins')
            ->set(['foo' => 'bar', 'users' => ['Jack', 'Kate', 'Sawyer', 'Lock', 'Jacob']])
            ->view('users');
```

If you need to append some value to a data you can use the `append` method:

```php
$this->slice->set(['foo' => 'bar', 'users' => ['Jack', 'Kate', 'Sawyer', 'Lock', 'Jacob']])
            ->append('users', 'Ben')
            ->view('users');
```

## Localization

Slice's localization features provide a convenient way to retrieve strings in various languages, allowing you to easily support multiple languages within your application.

If you want to support multiple languages in your application, you would provide folders inside your `application/language/` directory for each of them, and you would specify the default language in your `application/config/config.php`:

```plain
  application/
              language/
                      english/
                              error_lang.php
                              message_lang.php
                      french/
                              error_lang.php
                              message_lang.php
```

### Configuring The Locale

The default language for your application is stored in the `application/config/config.php` file. Of course, you may modify this value to suit the needs of your application. You may also change the active language using the `locale` method:

```php
$this->slice->locale('french');
```

### Retrieving Translation Strings

You may retrieve lines from language files using the `@lang` directive. The `@lang` accepts the file and key of the translation string as its first argument. For example, let's retrieve the welcome translation string from the `application/language/english/message_lang.php` language file:

```php
@lang('message.welcome')
```

If the specified translation string does not exist, the `@lang` directive will simply return the translation string key. So, using the example above, the `@lang` directive would return '*messages.welcome*' if the translation string does not exist.

### Replacing Parameters In Translation Strings

If you wish, you may define place-holders in your translation strings. All place-holders are prefixed with a `:`. For example, you may define a welcome message with a place-holder name:

```php
$lang['welcome'] = 'Welcome, :Name';
```

To replace the place-holders when retrieving a translation string, pass an array of replacements as the second argument to the `@lang` directive:

```php
@lang('messages.welcome', ['name' => 'GustMartins'])
```

If your place-holder contains all capital letters, or only has its first letter capitalized, the translated value will be capitalized accordingly:

```php
$lang['welcome'] = 'Welcome, :NAME';  //  Welcome, GUSTMARTINS
$lang['goodbye'] = 'Goodbye, :Name';  //  Goodbye, Gustmartins
```

### Pluralization

Pluralization is a complex problem, as different languages have a variety of complex rules for pluralization. By using a "pipe" character, you may distinguish singular and plural forms of a string:

```php
$lang['apples'] = 'There is one apple|There are many apples';
```

You may even create more complex pluralization rules which specify translation strings for multiple number ranges:

```php
$lang['apples'] = '{0} There are none|[1,19] There are some|[20,*] There are many';
```

After defining a translation string that has pluralization options, you may use the `@choice` directive to retrieve the line for a given "count". In this example, since the count is greater than one, the plural form of the translation string is returned:

```php
@choice('messages.apples', 10)
```

## Slice Syntax

Just like Laravel's Blade, Slice-Library is simple, and powerful! 
Slice-Library does not restrict you from using plain PHP code in your views. All  Slice views are compiled into plain PHP code and cached until they are modified, meaning Slice adds essentially zero overhead to your application. Slice view files use the `.slice.php` file extension, but you can change it in the `application/config/slice.php` file.

### Defining a Layout

To get started, let's take a look at a simple example. First, we will examine a "master" page layout. Since most web applications maintain the same general layout across various pages, it's convenient to define this layout as a single Slice view:

```HTML
<!-- Stored in application/views/layouts/app.slice.php -->
<html>
    <head>
        <title>App Name - @yield('title')</title>
    </head>
    <body>
        @section('sidebar')
            This is the master sidebar.
        @show

        <div class="container">
            @yield('content')
        </div>
    </body>
</html>
```

The `@section` directive, as the name implies, defines a section of content, while the `@yield` directive is used to display the contents of a given section.

### Extending A Layout

When defining a child view, use the Slice `@extends` directive to specify which layout the child view should "inherit".

```HTML
<!-- Stored in application/views/child.slice.php -->

@extends('layouts.app')

@section('title', 'Child Page Title')

@section('sidebar')
    @parent

    <p>This is appended to the master sidebar.</p>
@endsection

@section('content')
    <p>This is my body content.</p>
@endsection
```

In this example, the sidebar section is utilizing the `@parent` directive to append (rather than overwriting) content to the layout's sidebar. The `@parent` directive will be replaced by the content of the layout when the view is rendered.

Now, you just need to load the view with the Slice `view` method:

```php
$this->slice->view('child');
```

### Displaying Data

You may display data passed to your Slice views by wrapping the variable in curly braces. For example, you may display the contents of a `name` variable like so:

```html
Hello, {{ $name }}.
```

Of course, you are not limited to displaying the contents of the variables passed to the view. You may also echo the results of any PHP function as well as CodeIgniter loaded library's functions:

```html
Find out more in this {{ anchor('anchor/this', 'link') }}
```

### Echoing Data If It Exists

Sometimes you may wish to echo a variable, but you aren't sure if the variable has been set. But, instead of writing a ternary statement, Slice provides you with the following convenient shortcut, which will be compiled to the ternary statement:

```html
{{ $name or 'Default' }}
```

In this example, if the `$name` variable exists, its value will be displayed. However, if it does not exist, the word `Default` will be displayed.

### Slice & JavaScript Frameworks

Since many JavaScript frameworks also use "curly" braces to indicate a given expression should be displayed in the browser, you may use the `@` symbol to inform the Slice-Library rendering engine an expression should remain untouched. For example:

```html
<h1>Slice Library</h1>

Hello, @{{ username }}
```

In this example, the `@` symbol will be removed by Slice; however, `{{ username }}` expression will remain untouched by the Slice-Library engine, allowing it to instead be rendered by your JavaScript framework.

### If Statements

You may construct `if` statements using the `@if`, `@elseif`, `@else`, and `@endif` directives. These directives function identically to their PHP counterparts:

```php
@if (count($records) === 1)
    I have one record!
@elseif (count($records) > 1)
    I have multiple records!
@else
    I do not have any records!
@endif
```

Slice-Library also provides an `@unless` directive:

```php
@unless (count($users) != 0)
    There are no users to show.
@endunless
```

In addition to the conditional directives already discussed, you may use the `@isset` and `@empty` directives as convenient shortcuts for their respective PHP functions:

```php
@isset($rows)
  //  $rows is defined and is not null...
@endisset

@empty($rows)
  //  $rows is "empty"...
@endempty
```

### Loops

Slice-Library provides simple directives for working with PHP's loop structures. Again, each of these directives functions identically to their PHP counterparts:

```php
@for ($i = 0; $i < 10; $i++)
    The current value is {{ $i }}
@endfor

@foreach ($users as $key => $user)
    <p>This is user {{ $user }}</p>
@endforeach

@forelse ($users as $key => $user)
    <li>{{ $user }}</li>
@empty
    <p>No users</p>
@endforelse

@while (true)
    <p>I am looping forever.</p>
@endwhile
```

When using loops you may also end the loop or skip the current iteration:

```php
@foreach ($users as $key => $user)
    @if ($user == 'Kate')
        @continue
    @endif

    <li>{{ $user }}</li>

    @if ($key == 4)
        @break
    @endif
@endforeach
```

You may also include the condition with the directive declaration in one line:

```php
@foreach ($users as $key => $user)
    @continue($key == 1)

    <li>{{ $user }}</li>

    @break($key == 5)
@endforeach
```

### Comments

Slice-Libray also allows you to define comments in your views. However, unlike HTML comments, Slice comments are not included in the HTML returned by your application:

```php
{{-- This comment will not be present in the rendered HTML --}}
```

### PHP

In some situations, it's useful to embed PHP code into your views. You can use the Slice `@php` directive to execute a block of plain PHP within your template:

```php
@php
    //
@endphp
```

### Including Sub-Views

Slice's `@include` directive allows you to include a Slice view from within another view. All variables that are available to the parent view will be made available to the included view:

```php
<div>
    @include('public.errors')

    <form>
        <!-- Form Contents -->
    </form>
</div>
```

Even though the included view will inherit all data available in the parent view, you may also pass an array of extra data to the included view:

```php
@include('view.name', ['some' => 'data'])
```

Of course, if you attempt to `@include` a view which does not exist, Slice-Library will throw an error. If you would like to include a view that may or may not be present, you should use the @includeIf directive:

```php
@includeIf('view.name', ['some' => 'data'])
```

### Rendering Views For Collections

It is very useful to combine loops and includes into one line and it is possible with Slice's `@each` directive:

```php
@each('view.name', $cars, 'car')
```

The first argument is the view partial to render for each element in the array. The second argument is the array you wish to iterate over, while the third argument is the variable name that will be assigned to the current iteration within the view.

In the example, if you are iterating over an array of `cars`, typically you will want to access each car as a `car` variable within your view partial.

The key for the current iteration will be available as the `key` variable within your view partial.

You may also pass a fourth argument to the `@each` directive. This argument determines the view that will be rendered if the given array is empty.

```php
@each('view.name', $cars, 'car', 'view.empty')
```

### Extending Slice-Library Directives

Slice-Library allows you to define your own custom directives using the `directive` method. When the Slice compiler encounters the custom directive, it will call the provided callback with the expression that the directive contains.

The following example creates a @slice('Text') directive which echo a given string:

```php
<?php defined('BASEPATH') OR exit('No direct script access allowed');

class Test extends CI_Controller
{
   public function __construct()
   {
      parent::__construct();
      $this->load->library('slice');
   }

   public function index()
   {
      $this->slice->with('username', 'John Doe')
                  ->directive('Test::custom_slice_directive')
                  ->view('users.profile');
   }

   static function custom_slice_directive($content)
   {
      // Finds for @slice directive
      $pattern = '/(\s*)@slice\s*\((\'.*\')\)/';
      return preg_replace($pattern, '$1<?php echo "$2"; ?>', $content);
   }
}
```

## Contributions

This package was created by [Gustavo Martins][GustMartins], but your help is welcome! Things you are welcome to do:

+ Report any bug you may encounter
+ Suggest a feature for the project

For more information about contributing to the project please, read the [Contributing Requirements][contrib].

## Change Log

Currently, the Slice-Library is in the version **1.3.0**. See the full [Changelog][changelog] for more details.

[GustMartins]: https://github.com/GustMartins
[contrib]: https://github.com/GustMartins/Slice-Library/blob/master/contributing.md
[changelog]: https://github.com/GustMartins/Slice-Library/blob/master/changelog.md


changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.3.0] - 2017-09-07

### Added
+ Helper file with 89 functions!

### Changed
+ Rename _inflector() function to `inflector()` and make it public

## [1.2.1] - 2017-05-22

### Changed
+ Rename _i18n() function to `i18n()` and make it public

### Fixed
+ Bug when two or more `@lang()` or `@choice()` were placed in the same line

## [1.2.0] - 2017-05-17

### Added
+ 'Slice Execution Time' to benchmark section of the profiler
+ `@lang` and `@choice` directives
+ Autoload CodeIgniter libraries and helpers

### Fixed
+ Bug where config items were not initialized

## [1.1.0] - 2017-05-04

### Added
+ `@each` directive
+ `@isset` and `@empty` directives

## 1.0.2 - 2017-04-23

### Changed
+ README.md revised to solve some writing errors

### Fixed
+ Bug where empty lines were added in the beginning of a rendered view

## [1.0.1] - 2017-03-31

### Fixed
+ Bug when using Modular Extensions - HMVC

## [1.0.0] - 2017-02-23
First publicly released version.


[1.3.0]: https://github.com/GustMartins/Slice-Library/releases/tag/1.3.0
[1.2.1]: https://github.com/GustMartins/Slice-Library/releases/tag/1.2.1
[1.2.0]: https://github.com/GustMartins/Slice-Library/releases/tag/1.2.0
[1.1.0]: https://github.com/GustMartins/Slice-Library/releases/tag/1.1.0
[1.0.1]: https://github.com/GustMartins/Slice-Library/releases/tag/1.0.1
[1.0.0]: https://github.com/GustMartins/Slice-Library/releases/tag/1.0.0


application/controllers/Test.php
<?php
defined('BASEPATH') OR exit('No direct script access allowed');

/**
 *  This is a test class to show you how to work with Slice-Library!
 */
class Test extends CI_Controller
{
	public function __construct()
	{
		parent::__construct();
		//	Load the Slice-Library as any CodeIgniter library
		$this->load->library('slice');
		$this->load->helper('url');
	}

	public function index()
	{
		/**
		 *  The with() method adds a pair of key/value data.
		 *  The directive() method inserts a custom directive
		 *  The view() method shows the page located at 'application/views/page.slice.php'
		 */
		$this->slice->with('full_name', 'John Doe')
					->with('users', [['id' => 1, 'name' => 'John'], ['id' => 2, 'name' => 'Jane'], ['id' => 3, 'name' => 'Dave'], ['id' => 4, 'name' => 'Arthur'], ['id' => 5, 'name' => 'Michael'], ['id' => 6, 'name' => 'Ben']])
					->with('members', [])
					->with('rows', '0')
					->directive('Test::custom_slice_directive');

		//	Slice-Library 1.3 comes with helpers to make things easier!
		view('page');
	}

	/**
	 *  This is a custom function that adds a custom directive to Slice-Library!
	 *  This function finds for @slice('string') directive and shows the string
	 */
	static function custom_slice_directive($content)
	{
		$pattern = '/(\s*)@slice\s*\((\'.*\')\)/';
		return preg_replace($pattern, '$1<?php echo "$2"; ?>', $content);
	}
}


application/config/slice.php
<?php
defined('BASEPATH') OR exit('No direct script access allowed');

/**
 * @package		Slice
 * @author		Gustavo Martins <gustavo_martins92@hotmail.com>
 * @link		https://github.com/GustMartins/Slice-Library
 * @since		Version 1.0.0
 */

/*
|--------------------------------------------------------------------------
| Slice File Extension
|--------------------------------------------------------------------------
|
| Set the file extension for the slice template
|
*/
$config['slice_ext'] = '.slice.php';

/*
|--------------------------------------------------------------------------
| Cache Expiration Time
|--------------------------------------------------------------------------
|
| Set the amount of time to keep the file in cache
|
*/
$config['cache_time'] = 3600;

/*
|--------------------------------------------------------------------------
| Enable/Disable Autoload
|--------------------------------------------------------------------------
|
| Set to TRUE to autoload CodeIgniter Libraries and Helpers
|
*/
$config['enable_autoload'] = FALSE;

/*
|--------------------------------------------------------------------------
| Resources to Autoload
|--------------------------------------------------------------------------
|
| List of Libraries and Helpers to autoload with Slice-Library.
|
| WARNING: To autoload this resources you must set 'enable_autoload'
| variable to TRUE.
|
*/
$config['libraries'] = array();
$config['helpers'] = array();

/*
|--------------------------------------------------------------------------
| Load Slice Helper
|--------------------------------------------------------------------------
|
| Set to TRUE and Slice helper file will be loaded in the initialization.
| Make sure you have the proper file at application/helper/slice_helper.php
|
*/
$config['enable_helper'] = TRUE;


application/views/_css.slice.php
<style type="text/css">
	::selection { background-color: #E13300; color: white; }
	::-moz-selection { background-color: #E13300; color: white; }

	body {
		background-color: #fff;
		margin: 40px;
		font: 13px/20px normal Helvetica, Arial, sans-serif;
		color: #4F5155;
	}

	a {
		color: #003399;
		background-color: transparent;
		font-weight: normal;
	}

	h1 {
		color: #444;
		background-color: transparent;
		border-bottom: 1px solid #D0D0D0;
		font-size: 19px;
		font-weight: normal;
		margin: 0 0 14px 0;
		padding: 14px 15px 10px 15px;
	}

	code {
		font-family: Consolas, Monaco, Courier New, Courier, monospace;
		font-size: 12px;
		background-color: #f9f9f9;
		border: 1px solid #D0D0D0;
		color: #002166;
		display: block;
		margin: 14px 0 14px 0;
		padding: 12px 10px 12px 10px;
	}

	#body {
		margin: 0 15px 0 15px;
	}

	p.footer {
		text-align: right;
		font-size: 11px;
		border-top: 1px solid #D0D0D0;
		line-height: 32px;
		padding: 0 10px 0 10px;
		margin: 20px 0 0 0;
	}

	#container {
		margin: 10px;
		border: 1px solid #D0D0D0;
		box-shadow: 0 0 8px #D0D0D0;
	}
</style>


application/views/page.slice.php
@extends('index')

@section('title', 'Welcome to CodeIgniter - Page file extending index file')

@section('top_content')
	@parent

	<p>This paragraph was add in the page 'page.slice.php'</p>
@endsection

@section('content')
	<hr>

	<h3>If Statements:</h3>
	<code>&#64;if (count($users) === 1) <br>  &nbsp; I have one user! <br> &#64;elseif (count($users) > 1) <br>  &nbsp; I have multiple users! <br> &#64;else <br>  &nbsp; I don't have any user! <br> &#64;endif</code>
	<p>This example would result:</p>
	@if (count($users) === 1)
		<p>I have one user!</p>
	@elseif (count($users) > 1)
		<p>I have multiple users!</p>
	@else
		<p>I don't have any user!</p>
	@endif
	<br>
	<p>For convenience, Slice also provides an <mark>@unless</mark> directive:</p>
	<code>&#64;unless ($members) <br> &nbsp; You are not signed in. <br> &#64;endunless</code>
	<p>This example would result:</p>
	@unless ($members)
		There are no members at the moment.
	@endunless
	<br>
	<p>In addition to the conditional directives already discussed, you may use the <mark>&#64;isset</mark> and <mark>&#64;empty</mark> directives as convenient shortcuts for their respective PHP functions:</p>
	<code>&#64;isset($rows) <br>   //  $rows is defined and is not null... <br> &#64;endisset <br><br>&#64;empty($rows) <br>   //  $rows is "empty"... <br> &#64;endempty</code>

	<hr>

	<h3>Loopings:</h3>
	<code>&#64;for ($i = 0; $1 < 10; $i++) <br> &nbsp; The current value is &#123;&#123; $i }} <br> &#64;endfor</code>
	<p>This example would result:</p>
	<ul>
	@for ($i = 0; $i < 10; $i++)
		<li>The current value is {{ $i }}</li>
	@endfor
	</ul>
	<br>
	<code>&#64;foreach ($users as $user) <br> &nbsp; This is user &#123;&#123; $user['name'] }} <br> &#64;endforeach</code>
	<p>This example would result:</p>
	<ul>
		@foreach ($users as $user)
			<li>This is user {{ $user['name'] }}</li>
		@endforeach
	</ul>
	<br>
	<code>&#64;forelse ($members as $member) <br> &nbsp;&nbsp; &#123;&#123; $member['name'] }} <br> &#64;empty <br> &nbsp;&nbsp; No members <br> &#64;endforelse<br> </code>
	<p>This example would result:</p>
	<ul>
		@forelse ($members as $member)
			<li>{{ $member['name'] }}</li>
		@empty
			<li>No members</li>
		@endforelse
	</ul>
	<br>
	<p>When using loops you may also end the loop or skip the current iteration:</p>
	<code>&#64;foreach ($users as $user) <br> &nbsp; &#64;if ($user['id'] == 1) <br> &nbsp; &nbsp; &#64;continue <br> &nbsp; &#64;endif <br><br> &nbsp; &nbsp; &#123;&#123; $user['name'] }} <br><br> &nbsp; &#64;if ($user['id'] == 5) <br> &nbsp; &nbsp; &#64;break <br> &nbsp; &#64;endif <br> &#64;endforeach</code>
	<p>This example would result:</p>
	<ul>
		@foreach ($users as $user)
			@if ($user['id'] == 1)
				@continue
			@endif

			<li>{{ $user['name'] }} - {{ $user['id'] }}</li>

			@if ($user['id'] == 5)
				@break
			@endif
		@endforeach
	</ul>
	<p>You may also include the condition with the directive declaration in one line:</p>
	<code>&#64;foreach ($users as $user) <br>  &nbsp;  &nbsp; &#64;continue($user['id'] == 2) <br> <br> &nbsp; &nbsp; &#123;&#123; $user['name'] }} - &#123;&#123; $user['id'] }} <br> <br> &nbsp; &nbsp; &#64;break($user['id'] == 4) <br> &#64;endforeach</code>
	<p>This example would result:</p>
	<ul>
		@foreach ($users as $user)
			@continue($user['id'] == 2)

			<li>{{ $user['name'] }} - {{ $user['id'] }}</li>

			@break($user['id'] == 4)
		@endforeach
	</ul>

	<hr>

	<h3>Setup your own directives:</h3>
	<p>Slice allows you to define your own custom directives using the <mark>directive</mark> method. When the Slice compiler encounters the custom directive, it will call the provided callback with the expression that the directive contains.</p>
	<p>The following example creates a <mark>@slice($var)</mark> directive that only display the text of the $var. This is a method in the controller of the CodeIgniter:</p>
	<code>static function custom_slice_directive($content) <br> { <br>  &nbsp; &nbsp; //	Finds for @slice directive <br>  &nbsp; &nbsp; $pattern = '/(\s*)@slice\s*\((\'.*\')\)/'; <br>  &nbsp; &nbsp; return preg_replace($pattern, '$1<?php echo "$2"; ?>', $content); <br> } </code>
	<p>Then you have to use it with the method <mark>directive</mark>, like this:</p>
	<code>$this->slice->directive('Test::custom_slice_directive');</code>
@endsection


application/views/index.slice.php
@php
	defined('BASEPATH') OR exit('No direct script access allowed');
@endphp

<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="utf-8">
	<title>@yield('title', 'Welcome to CodeIgniter')</title>

	@include('_css')
	@includeIf('_js')
</head>
<body>

<div id="container">
	@yield('page_title', '<h1>Welcome to CodeIgniter Slice-Library</h1>')

	<div id="body">
		@section('top_content')
			<p>This {{ anchor('', 'paragraph') }} belongs to a section called 'top content'! It must always be shown! Sometimes other pages will add another paragraph to this section. <br> You can open a section like this:</p>
			<code>&#64;section('top_section')</code>
			<p>You need to close the section with the directive:</p>
			<code>&#64;endsection</code>
			<p>If you want to close the section and display it, use this directive:</p>
			<code>&#64;show</code>
		@show

		<hr>

		<h3>Comments:</h3>
		<p>Slice allows you to define comments in your views. However, unlike HTML comments, Blade comments are not included in the HTML returned by your application. <br> There is a PHP comment between this paragraph and the code example bellow.</p>
		{{-- This is a Slice comment --}}
		<code>@{{&#45;&#45; This is a Slice comment --}} </code>

		<br><hr>

		<h3>PHP Tags:</h3>
		<p>In some situations, it's useful to embed PHP code into your views. You can use the Slice <mark>&#64;php</mark> directive to execute a block of plain PHP within your template. <br> In this page you can see an example in the begging of the page.</p>
		<code>&#64;php <br> // <br> &#64;endphp</code>

		<br><hr>

		<h3>Echoing variables:</h3>
		<p>You may display data passed to your Slice views by wrapping the variable in curly braces. You may display the contents of the <mark>full_name</mark> variable like so:</p>
		<code>&#123;&#123; $full_name }}</code>
		<p>The code above would return: <strong>{{ $full_name }}</strong></p>
		<br>
		<h3>Echoing Data If It Exists:</h3>
		<p>Sometimes you may wish to echo a variable, but you aren't sure if the variable has been set. However, instead of writing a ternary statement, Slice provides you with the following convenient shortcut, which will be compiled to the ternary statement:</p>
		<code>&#123;&#123; $name or 'Default' }}</code>
		<p>In this example, if the <mark>$name</mark> variable exists, its value will be displayed. However, if it does not exist, the word <strong>Default</strong> will be displayed.</p>
		<p>The code above would return: <strong>{{ $name or 'Default' }}</strong></p>

		<br><hr>

		<h3>Preserving braces:</h3>
		<p>Since many JavaScript frameworks also use "curly" braces to indicate a given expression should be displayed in the browser, you may use the <mark>@</mark> symbol to inform the Slice rendering engine an expression should remain untouched. For example:</p>
		<code>Hello, @&#123;&#123; name }}</code>
		<p>In this example, the @ symbol will be removed by Blade; however, &#123;&#123; name }} expression will remain untouched by the Blade engine, allowing it to instead be rendered by your JavaScript framework.</p>
		<p>The code above would return: <strong>Hello, @{{ name }}</strong></p>

		@yield('content')

		@section('bottom_content')
			<p>This paragraph is the last one!</p>
		@show
	</div>

	<p class="footer">Page rendered in <strong>{elapsed_time}</strong> seconds. <?php echo  (ENVIRONMENT === 'development') ?  'CodeIgniter Version <strong>' . CI_VERSION . '</strong>' : '' ?></p>
</div>

</body>
</html>
